lib/faraday/request/retry.rb

```diff
@@ -106,26 +112,32 @@ def initialize(app, options = nil)
       @errmatch = build_exception_matcher(@options.exceptions)
     end
 
-    def sleep_amount(retries)
+    def sleep_amount(retries, env)
       retry_index = @options.max - retries
       current_interval = @options.interval * (@options.backoff_factor ** retry_index)
       current_interval = [current_interval, @options.max_interval].min
       random_interval  = rand * @options.interval_randomness.to_f * @options.interval
-      current_interval + random_interval
+
+      retry_after = calculate_retry_after(env)
+      interval    = current_interval + random_interval
+
+      retry_after && retry_after >= interval ? retry_after : interval
```

Comment:
This method looks a bit too complicated now, so I would separate the logic to calculate the retry interval into a separate method (similarly to what you did with `calculate_retry_after`).
This new method could be called `calculate_retry_interval(retries)`.